delivery_location/tests/test_deliverylocation.py

```diff
@@ -0,0 +1,112 @@
+
+
+@pytest.fixture
+def delivery_location1(supplier0):
+    return DeliveryLocation(user_name=supplier0, location="Haifa", date=datetime.date(2022, 12, 31))
+
+
+class TestDeliveryLocationModel:
+    @pytest.mark.django_db()
+    def test_add_delivery_location(self, supplier0, delivery_location0, delivery_location1):
+        supplier0.save()
+        delivery_location0.add_delivery_location()
+        delivery_location1.add_delivery_location()
+        assert delivery_location0 in DeliveryLocation.objects.all()
+        assert delivery_location1 in DeliveryLocation.objects.all()
+
+    @pytest.mark.django_db()
+    def test_delete_non_existing_delivery_location(self):
+        with pytest.raises(DeliveryLocation.DoesNotExist):
+            DeliveryLocation.remove_delivery_location_by_id(-1)
+
+    @pytest.mark.django_db()
+    def test_delete_delivery_location(self, supplier0, delivery_location0):
+        supplier0.save()
+        delivery_location0.add_delivery_location()
+        assert delivery_location0 in DeliveryLocation.objects.all()
+        delivery_location0.remove_delivery_location()
+        assert delivery_location0 not in DeliveryLocation.objects.all()
+
+    @pytest.mark.django_db()
+    def test_filter_by_id(self, supplier0, delivery_location0):
+        supplier0.save()
+        delivery_location0.add_delivery_location()
+        assert delivery_location0 == DeliveryLocation.get_delivery_location_by_id(delivery_location0.id)
+        with pytest.raises(DeliveryLocation.DoesNotExist):
+            DeliveryLocation.get_delivery_location_by_id(-1)
+
+    @pytest.mark.django_db()
+    def test_filter_by_id_that_does_not_exist(self, supplier0, delivery_location0):
+        supplier0.save()
+        delivery_location0.add_delivery_location()
+        with pytest.raises(DeliveryLocation.DoesNotExist):
+            DeliveryLocation.get_delivery_location_by_id(-1)
+
+    @pytest.mark.django_db()
+    def test_filter_by_location_from_supplier(self, supplier0, delivery_location0):
+        supplier0.save()
+        delivery_location0.add_delivery_location()
+        assert delivery_location0 in list(
+            DeliveryLocation.filter_by_supplier_and_location(supplier0, "Qiryat Shemona"))
+
+    @pytest.mark.django_db()
+    def test_filter_by_location_that_does_not_exist_from_supplier(self, supplier0):
+        supplier0.save()
+        assert [] == list(DeliveryLocation.filter_by_supplier_and_location(supplier0,
+                                                                           "LOCATION THAT IS NOT RELATED TO SUPPLIER0"))
+
+    @pytest.mark.django_db()
+    def test_filter_by_location(self, supplier0, delivery_location0, delivery_location1):
+        supplier0.save()
+        delivery_location0.add_delivery_location()
+        delivery_location1.add_delivery_location()
+        assert [delivery_location0, delivery_location1] == list(
```

Comment:
I'm not sure whether we can tell for sure `filter_by_location` will always return the same order of locations. If not, this test can become flaky (test with unstable results). I'd suggest sorting the result:

```suggestion
        assert sorted([delivery_location0, delivery_location1]) == sorted(list(DeliveryLocation.filter_by_location(supplier0)))
```
But it might be that those objects don't have natural order according to which they can be sorted, so if this fails, you should be able to do something like this:

```suggestion
        assert sorted([delivery_location0, delivery_location1], key=str) == sorted(list(DeliveryLocation.filter_by_location(supplier0)), key=str)
```

Another option will be to check that both values are in the returned list + checking the length of the returned list. Something like:
```
output = list(DeliveryLocation.filter_by_location(supplier0))
locations = [delivery_location0, delivery_location1]
assert all(location in locations for location in output)
assert len(output) == len(locations)
```

Reply:
You are completely right,i need to sort them in order to check that both of them identical,thanks!
fixed

Reply:
Or we can use assert sorted([delivery_location0, delivery_location1], key=lambda x:x.id) == sorted(list(DeliveryLocation.filter_by_location(supplier0)), key=x:x.id)
Which one do you prefer other than the str.

Reply:
Whatever you prefer.

Another option which might be even cleaner is to compare them as sets. The order in sets doesn't matter, but they include no duplicate instances.

If your function could have returned duplicate instances, then this would not have been a good use of sets, but in this case, I don't think you can get duplicate instances from your database, so I think that should be ok.

```
In [18]: set([1,2,3,3])
Out[18]: {1, 2, 3}

In [19]: [1,2,3] == [3,2,1]
Out[19]: False

In [20]: set([1,2,3]) == set([3,2,1])
Out[20]: True
```

So, if you call `set()` on both sides of this comparison, I think it will still be legitimate comparison and cleaner than the other approaches.

Reply:
Thanks i didnt know that comparision betwen sets doesnt matter but i didnt know that if the order is not the same with python then the same case happends, thanks new things to use in the future,fixed to be set since it is the right case.
Fixed!

Reply:
Cool. Sets are convenient for making stable tests. Just remember not to use them if you care about repetitions or order of elements.

Reply:
Understood!